core-foundation/src/array.rs

```diff
@@ -12,36 +12,51 @@
 pub use core_foundation_sys::array::*;
 pub use core_foundation_sys::base::{CFIndex, CFRelease};
 use core_foundation_sys::base::{CFTypeRef, kCFAllocatorDefault};
-use base::CFType;
 use libc::c_void;
 use std::mem;
 use std::marker::PhantomData;
+use std;
+use std::ops::Deref;
+use std::fmt::{Debug, Formatter};
 
-use base::{CFIndexConvertible, TCFType, CFRange};
+use base::{CFIndexConvertible, TCFType, TCFTypeRef, CFRange};
 
 /// A heterogeneous immutable array.
 pub struct CFArray<T = *const c_void>(CFArrayRef, PhantomData<T>);
 
-/// A trait describing how to convert from the stored *const c_void to the desired T
-pub unsafe trait FromVoid {
-    unsafe fn from_void(x: *const c_void) -> Self;
+/// A reference to an element inside the array
+pub struct ItemRef<'a, T: 'a>(T, PhantomData<&'a T>);
+
+impl<'a, T> Deref for ItemRef<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.0
+    }
 }
 
-unsafe impl FromVoid for u32 {
-    unsafe fn from_void(x: *const c_void) -> u32 {
-        x as usize as u32
+impl<'a, T: Debug> Debug for ItemRef<'a, T> {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), std::fmt::Error> {
+        self.0.fmt(f)
     }
 }
 
-unsafe impl FromVoid for *const c_void {
-    unsafe fn from_void(x: *const c_void) -> *const c_void {
-        x
+/// A trait describing how to convert from the stored *const c_void to the desired T
+pub unsafe trait FromVoid {
+    unsafe fn from_void<'a>(x: *const c_void) -> ItemRef<'a, Self> where Self: std::marker::Sized;
+}
+
+unsafe impl FromVoid for u32 {
+    unsafe fn from_void<'a>(x: *const c_void) -> ItemRef<'a, Self> {
+        // Functions like CGFontCopyTableTags treat the void*'s as u32's
+        // so we convert by casting directly
+        ItemRef(x as u32, PhantomData)
     }
 }
 
-unsafe impl FromVoid for CFType {
-    unsafe fn from_void(x: *const c_void) -> CFType {
-        TCFType::wrap_under_get_rule(mem::transmute(x))
+unsafe impl<T: TCFType> FromVoid for T {
+    unsafe fn from_void<'a>(x: *const c_void) -> ItemRef<'a, Self> {
+        ItemRef(TCFType::wrap_under_create_rule(T::Ref::from_void_ptr(x)), PhantomData)
```

Comment:
Won't the drop of the constructed `T` do a `CFRelease` and make the retain count wrong? If you construct a new `T` from a void pointer, should you not do it under the get rule?

Could you add a test that checks retain counts are correct before and after borrowing values from the array?

Also, nitpicking: `T::wrap_under_...` should work as well? Making it a bit more clear which type we create here.

Reply:
Yep. I've messed up the reference counts here. I need to think about how to fix it. wraping with the get rule will increase the retain count which defeats the point of this patch.

Reply:
The underlying ffi function, `CFArrayGetValueAtIndex` specifies that values returned should follow the get rule. So if you would use `CFArray` directly in Objective-C you would likely bump the retain count? So is there any reason we should be allowed not to? If we convert the value into a CF object and hold a reference to that then should the retain count not reflect how many handles there are to that object?

Reply:
Rust, in theory, should let us use the existing reference in the array safely. Since its lifetime will be tied to the array we should be able to interact with those objects without needing to bump the retain count.

i.e. imagine if you have Vec<Rc<u32>>. You can iterate over them without having to bump the reference counts on all of the objects.

Reply:
Yes true. As long as the borrow can't outlive the array yes indeed!

How about you just `mem::forget` the `T` value in the `Drop` impl for `ItemRef`?

Reply:
fn drop takes &mut self so it won't let me just mem::forget() the T value

Reply:
But using ManuallyDrop let's me do it.